### PR TITLE
collation: cast charset according to the function's resulting charset

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/28356

some functions like `concat`, `eq` may have different charset among the args. we will infer the charset and collation according to the args, so, if the resulting charset is different from the arg's charset, we need to cast the arg's charset to the resulting charset. e.g. 
```sql
select concat(a, 0x31) from t;
```
if a is `gbk` charset, we should convert `0x31` to gbk charset. If this convert is impossible, for example, 0x81 is not a valid gbk character, we will return an error.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

**Review by Defined2014 — APPROVED** [review]

File: types/datum.go [hunk 1]
```diff
@@ -996,9 +996,9 @@ func ProduceStrWithSpecifiedTp(s string, tp *FieldType, sc *stmtctx.StatementCon
 		// overflowed part is all whitespaces
 		var overflowed string
 		var characterLen int
-		// Flen is the rune length, not binary length, for UTF8 charset, we need to calculate the
+		// Flen is the rune length, not binary length, for Non-binary charset, we need to calculate the
 		// rune count and truncate to Flen runes if it is too long.
-		if chs == charset.CharsetUTF8 || chs == charset.CharsetUTF8MB4 {
+		if chs != charset.CharsetBinary {
```
Comment (tangenta) [inline types/datum.go:1001]:
```
mysql> select cast('😂' as char(10) charset utf8);

+---------------------------------------+
| cast('😂' as char(10) charset utf8)     |
+---------------------------------------+
| 😂                                      |
+---------------------------------------+
```

We may also need to handle the "mb3" case.

File: expression/collation.go [hunk 2]
```diff
@@ -322,7 +317,11 @@ func safeConvert(ctx sessionctx.Context, ec *ExprCollation, args ...Expression)
 			if err != nil {
 				return false
 			}
-			if !isNull && !isValidString(str, ec.Charset) {
+			// if value is NULL or binary string, just skip it.
+			if isNull || types.IsBinaryStr(c.GetType()) {
+				continue
+			}
```
Comment (tangenta) [inline expression/collation.go:323]: We can move `types.IsBinaryStr(c.GetType())` to the beginning of this loop to avoid unnecessary `EvalString`.

Comment (zimulala) [reply expression/collation.go:323]: @xiongjiwei Please address this comment.

**Comment by tangenta** [comment]
Close because there is another implementation: https://github.com/pingcap/tidb/pull/29905

**Pull request closed without merging**